flash/core/data/new_data_module.py

```diff
@@ -58,14 +85,19 @@ class DataModule(DataModule):
 
     input_transform_cls = DefaultInputTransform
     output_transform_cls = OutputTransform
-    flash_datasets_registry = FlashRegistry("datasets")
+    inputs_registry = FlashRegistry("datasets")
+    input_transforms_registry: Optional[FlashRegistry] = None
 
     def __init__(
         self,
-        train_dataset: Optional[Input] = None,
-        val_dataset: Optional[Input] = None,
-        test_dataset: Optional[Input] = None,
-        predict_dataset: Optional[Input] = None,
+        train_dataset: Optional[Dataset] = None,
+        val_dataset: Optional[Dataset] = None,
+        test_dataset: Optional[Dataset] = None,
+        predict_dataset: Optional[Dataset] = None,
+        train_transform: Optional[INPUT_TRANSFORM_TYPE] = None,
+        val_transform: Optional[INPUT_TRANSFORM_TYPE] = None,
+        test_transform: Optional[INPUT_TRANSFORM_TYPE] = None,
+        predict_transform: Optional[INPUT_TRANSFORM_TYPE] = None,
```

Comment:
This is quite confusing to me. Why do we need two ways to provide transforms?

Reply:
Hey @ethanwharris,

I believe the core question is: Should we support dataset being pyre pytorch dataset?

If the answer is no, then we can remove those transforms.
Otherwise, we would need to support the transform in order to simplify onboarding.

IMO, I would prefer to support pure PyTorch dataset there and `from_datasets` transform the sample into dictionaries which might not be the user desired behaviour.

Reply:
And if a user provides both an input with transform and a transform, it would raise a MisConfigurationExecption, so there is still 1 way to do something:

Use Input -> Transform are owned by the Input
Use PyTorch Dataset -> You can pass InputTransform as a utility method.

Reply:
Ok, you could do this now:

```py
    dm = DataModule.from_datasets(
        train_dataset=RandomDataset(64, 32),
        val_dataset=RandomDataset(64, 32),
        test_dataset=RandomDataset(64, 32),
        train_transform=TrainInputTransform,
        val_transform=_add_hundred,
        input_cls=Input,
        batch_size=3,
    )
    batch = next(iter(dm.train_dataloader()))
    assert batch[0][0] == 2
    batch = next(iter(dm.val_dataloader()))
    assert batch[0][0] == 101
    batch = next(iter(dm.test_dataloader()))
    assert batch[0][0] == 1
```

Let me remove support for it then.

Reply:
> Otherwise, we would need to support the transform in order to simplify onboarding.

With a standard PyTorch Dataset you would usually apply your transforms inside the `__getitem__`. So users would need to change their code here in order to provide the transforms this way.

> and from_datasets transform the sample into dictionaries which might not be the user desired behaviour.

Do you have an example of a case where that is undesired? All of our tasks require dictionaries so this seems like quite a big step back in simplicity in order to support a way of loading data that wouldn't work with any of our tasks.

Personally, I would rather remove the automatic conversion of samples to a dictionary in the DatasetInput. That way if a user wants to bring their own data set from PyTorch they would always be responsible for formatting their samples correctly and can have non-dictionary samples if that is what they want.

Reply:
Done !